File: simple-stack-example-conductor/src/main/java/com/zhuinden/navigationexamplecond/BaseKey.java
package com.zhuinden.navigationexamplecond;

import android.os.Bundle;
import android.os.Parcelable;

/**
 * Created by Owner on 2017. 06. 29..
 */

public abstract class BaseKey implements Parcelable {
    public final BaseController newController() {
        BaseController controller = createController();
        Bundle bundle = controller.getArgs();
        bundle.putParcelable("KEY", this);
        return controller;
    }

    protected abstract BaseController createController();
}

File: simple-stack-example-scoping/src/main/kotlin/com/zhuinden/simplestackexamplescoping/BaseKey.kt
package com.zhuinden.simplestackexamplescoping

import android.os.Bundle
import android.os.Parcelable

/**
 * Created by Zhuinden on 2018.09.17.
 */
interface BaseKey : Parcelable {
    val fragmentTag: String
        get() = toString()

    fun newInstance(): BaseFragment = createFragment().apply {
        arguments = (arguments ?: Bundle()).also { bundle ->
            bundle.putParcelable("KEY", this@BaseKey)
        }
    }

    fun createFragment(): BaseFragment
}
